File: DiabelskaKolejka.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>


#define N 10
#define P 4

pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t wagonReady=PTHREAD_COND_INITIALIZER;

int pasazerowieNaPokladzie=0;

void* pasazer(void* arg){
  int id=*((int*)arg);
  free(arg);


  printf("Pasażer %d próbuje wejść do wagonika...\n",id);

  pthread_mutex_lock(&mutex);

 while (pasazerowieNaPokladzie>=P){
  pthread_cond_wait(&wagonReady,&mutex);
 }
pasazerowieNaPokladzie++;

printf("Pasażer %d wszedł do wagonika. Liczba pasażerów: %d\n",id,pasazerowieNaPokladzie);

if(pasazerowieNaPokladzie==P){
printf("Wagonik rusza z %d pasażerami",P);
sleep(2);
printf("Wagonik zakończył przejazd!\n");

pasazerowieNaPokladzie=0;

pthread_cond_broadcast(&wagonReady);

}
pthread_mutex_unlock(&mutex);
pthread_exit(NULL);
}

int main(){
  pthread_t threads[N];
  for(int i=0;i<N;i++){
    int* id=malloc(sizeof(int));
    *id=i+1;
    pthread_create(&threads[i],NULL,pasazer,id);
    usleep(1000);
  }

  for(int i=0;i<N;i++){
    pthread_join(threads[i],NULL);
  }

  pthread_mutex_destroy(&mutex);
  pthread_cond_destroy(&wagonReady);

  return 0;

}

File: buforowanie/src/diabelskaKolejka.java
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class diabelskaKolejka {
    public static final int N = 10;
    public static final int P = 4;

    private final Lock mutex = new ReentrantLock();
    private final Condition condition = mutex.newCondition();

    private int pasazerowieNaPokladzie = 0;
    private int pasazerowieWysiedli = 0;

    public void pasazer(int id) {
        System.out.println("Pasażer " + id + " próbuje wejść do wagonika");

        mutex.lock();
        try {
            while (pasazerowieNaPokladzie >= P) {
                condition.await();
            }

            pasazerowieNaPokladzie++;
            System.out.println("Pasażer " + id + " wszedł do wagonika. Liczba pasażerów: " + pasazerowieNaPokladzie);

            if (pasazerowieNaPokladzie == P) {
                System.out.println("Wagonik rusza z " + P + " pasażerami...");

                try {
                    Thread.sleep(2000);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }

                System.out.println("Wagonik zakończył przejazd.");
                pasazerowieWysiedli = 0;
                condition.signalAll();
            } else {
                while (pasazerowieNaPokladzie != P) {
                    condition.await();
                }
            }

            pasazerowieWysiedli++;
            System.out.println("Pasażer " + id + " wyszedł z wagonika.");

            if (pasazerowieWysiedli == P) {
                pasazerowieNaPokladzie = 0;
                condition.signalAll();
            }

        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        } finally {
            mutex.unlock();
        }
    }

    public static void main(String[] args) {
        diabelskaKolejka kolejka = new diabelskaKolejka();
        for (int i = 0; i < N; i++) {
            final int id = i;
            new Thread(() -> kolejka.pasazer(id)).start();
        }
    }
}





